File: utilidad/rut_validator.py
"""
VALIDADOR DE RUT CHILENO - VERSIÓN MEJORADA
Sistema completo de validación, formateo y separación de RUT
Compatible con Django y uso independiente
"""

import re
from django.core.exceptions import ValidationError
from typing import Tuple, Dict


class RutValidator:
    """Clase para validar y manipular RUTs chilenos"""
    
    @staticmethod
    def limpiar(rut: str) -> str:
        """
        Limpia el RUT eliminando puntos, guiones y espacios.
        Convierte a mayúsculas.
        
        Args:
            rut: RUT a limpiar
            
        Returns:
            RUT limpio en mayúsculas
            
        Ejemplo:
            >>> RutValidator.limpiar('12.345.678-9')
            '123456789'
        """
        if not rut:
            return ''
        
        return str(rut).replace('.', '').replace('-', '').replace(' ', '').upper()
    
    @staticmethod
    def separar(rut: str) -> Dict[str, str]:
        """
        Separa el RUT en cuerpo y dígito verificador.
        
        Args:
            rut: RUT completo
            
        Returns:
            Diccionario con 'cuerpo' y 'dv'
            
        Ejemplo:
            >>> RutValidator.separar('12345678-9')
            {'cuerpo': '12345678', 'dv': '9'}
        """
        rut_limpio = RutValidator.limpiar(rut)
        
        if len(rut_limpio) < 2:
            return {'cuerpo': '', 'dv': ''}
        
        return {
            'cuerpo': rut_limpio[:-1],
            'dv': rut_limpio[-1]
        }
    
    @staticmethod
    def calcular_dv(cuerpo: str) -> str:
        """
        Calcula el dígito verificador de un RUT.
        
        Args:
            cuerpo: Cuerpo del RUT (sin DV)
            
        Returns:
            Dígito verificador calculado ('0'-'9' o 'K')
            
        Ejemplo:
            >>> RutValidator.calcular_dv('12345678')
            '5'
        """
        # Limpiar el cuerpo (solo números)
        cuerpo_limpio = re.sub(r'\D', '', str(cuerpo))
        
        if not cuerpo_limpio:
            return ''
        
        suma = 0
        multiplicador = 2
        
        # Recorrer de derecha a izquierda
        for digito in reversed(cuerpo_limpio):
            suma += int(digito) * multiplicador
            multiplicador = 7 if multiplicador == 7 else multiplicador + 1
            if multiplicador > 7:
                multiplicador = 2
        
        resto = suma % 11
        dv_calculado = 11 - resto
        
        # Retornar el DV correspondiente
        if dv_calculado == 11:
            return '0'
        elif dv_calculado == 10:
            return 'K'
        else:
            return str(dv_calculado)
    
    @staticmethod
    def validar(rut: str) -> bool:
        """
        Valida si un RUT es correcto.
        
        Args:
            rut: RUT completo a validar
            
        Returns:
            True si es válido, False en caso contrario
            
        Ejemplo:
            >>> RutValidator.validar('12.345.678-5')
            True
        """
        rut_limpio = RutValidator.limpiar(rut)
        
        # Validar longitud mínima
        if len(rut_limpio) < 2:
            return False
        
        # Validar formato: 7-8 dígitos + dígito verificador
        if not re.match(r'^\d{7,8}[0-9Kk]$', rut_limpio):
            return False
        
        datos = RutValidator.separar(rut_limpio)
        dv_calculado = RutValidator.calcular_dv(datos['cuerpo'])
        
        return datos['dv'] == dv_calculado
    
    @staticmethod
    def formatear(rut: str) -> str:
        """
        Formatea un RUT con puntos y guión (12.345.678-9).
        
        Args:
            rut: RUT a formatear
            
        Returns:
            RUT formateado
            
        Ejemplo:
            >>> RutValidator.formatear('123456789')
            '12.345.678-9'
        """
        datos = RutValidator.separar(rut)
        
        if not datos['cuerpo'] or not datos['dv']:
            return rut
        
        # Agregar puntos cada 3 dígitos de derecha a izquierda
        cuerpo = datos['cuerpo']
        cuerpo_formateado = ''
        contador = 0
        
        for i in range(len(cuerpo) - 1, -1, -1):
            if contador == 3:
                cuerpo_formateado = '.' + cuerpo_formateado
                contador = 0
            cuerpo_formateado = cuerpo[i] + cuerpo_formateado
            contador += 1
        
        return f"{cuerpo_formateado}-{datos['dv']}"
    
    @staticmethod
    def normalizar(rut: str) -> str:
        """
        Normaliza el RUT al formato sin puntos pero con guión (12345678-9).
        Este es el formato que se guarda en la base de datos.
        
        Args:
            rut: RUT a normalizar
            
        Returns:
            RUT normalizado
            
        Ejemplo:
            >>> RutValidator.normalizar('12.345.678-9')
            '12345678-9'
        """
        datos = RutValidator.separar(rut)
        
        if not datos['cuerpo'] or not datos['dv']:
            return rut
        
        return f"{datos['cuerpo']}-{datos['dv']}"


# ============================================
# FUNCIONES AUXILIARES PARA DJANGO
# ============================================

def validar_rut_chileno(value: str) -> str:
    """
    Validador de Django para campo RUT.
    Lanza ValidationError si el RUT no es válido.
    
    Args:
        value: RUT a validar
        
    Returns:
        RUT normalizado
        
    Raises:
        ValidationError: Si el RUT no es válido
    """
    if not value:
        raise ValidationError('El RUT es obligatorio.')
    
    rut_limpio = RutValidator.limpiar(value)
    
    # Validar formato
    if not re.match(r'^\d{7,8}[0-9Kk]$', rut_limpio):
        raise ValidationError(
            'Formato de RUT inválido. Use el formato: 12345678-9'
        )
    
    # Validar dígito verificador
    if not RutValidator.validar(value):
        raise ValidationError(
            'El dígito verificador del RUT es incorrecto.'
        )
    
    # Retornar normalizado
    return RutValidator.normalizar(value)


def normalizar_rut(rut: str) -> str:
    """
    Alias de RutValidator.normalizar para compatibilidad.
    
    Args:
        rut: RUT a normalizar
        
    Returns:
        RUT normalizado (formato: 12345678-9)
    """
    return RutValidator.normalizar(rut)


def validar_rut(value: str) -> str:
    """
    Alias de validar_rut_chileno para compatibilidad.
    """
    return validar_rut_chileno(value)


# ============================================
# GENERADOR DE RUT ALEATORIO (ÚTIL PARA TESTING)
# ============================================

def generar_rut_aleatorio() -> str:
    """
    Genera un RUT chileno válido aleatorio.
    Útil para pruebas y testing.
    
    Returns:
        RUT válido formateado
        
    Ejemplo:
        >>> rut = generar_rut_aleatorio()
        >>> RutValidator.validar(rut)
        True
    """
    import random
    
    # Generar cuerpo aleatorio (entre 1.000.000 y 99.999.999)
    cuerpo = random.randint(1000000, 99999999)
    dv = RutValidator.calcular_dv(str(cuerpo))
    
    return RutValidator.formatear(f"{cuerpo}{dv}")


# ============================================
# EJEMPLOS DE USO
# ============================================

if __name__ == '__main__':
    print("=== EJEMPLOS DE USO DEL VALIDADOR DE RUT ===\n")
    
    # Ejemplo 1: Validar RUT
    rut_test = '12.345.678-5'
    print(f"1. Validar RUT: {rut_test}")
    print(f"   ¿Es válido? {RutValidator.validar(rut_test)}\n")
    
    # Ejemplo 2: Separar RUT
    print(f"2. Separar RUT: {rut_test}")
    datos = RutValidator.separar(rut_test)
    print(f"   Cuerpo: {datos['cuerpo']}")
    print(f"   DV: {datos['dv']}\n")
    
    # Ejemplo 3: Calcular DV
    cuerpo = '12345678'
    print(f"3. Calcular DV para: {cuerpo}")
    dv = RutValidator.calcular_dv(cuerpo)
    print(f"   DV calculado: {dv}\n")
    
    # Ejemplo 4: Formatear RUT
    rut_sin_formato = '123456785'
    print(f"4. Formatear RUT: {rut_sin_formato}")
    print(f"   Formateado: {RutValidator.formatear(rut_sin_formato)}\n")
    
    # Ejemplo 5: Normalizar RUT
    rut_con_puntos = '12.345.678-5'
    print(f"5. Normalizar RUT: {rut_con_puntos}")
    print(f"   Normalizado: {RutValidator.normalizar(rut_con_puntos)}\n")
    
    # Ejemplo 6: Generar RUT aleatorio
    print("6. Generar RUT aleatorio:")
    rut_aleatorio = generar_rut_aleatorio()
    print(f"   RUT generado: {rut_aleatorio}")
    print(f"   ¿Es válido? {RutValidator.validar(rut_aleatorio)}\n")

File: utilidad/rut_validator.js
/**
 * VALIDADOR DE RUT CHILENO
 * Sistema completo de validación y formateo de RUT
 * Compatible con todos los formularios del proyecto
 */

class RutValidator {
    /**
     * Limpia el RUT eliminando puntos, guiones y espacios
     * @param {string} rut - RUT a limpiar
     * @returns {string} - RUT limpio
     */
    static limpiar(rut) {
        if (!rut) return '';
        return rut.toString()
            .replace(/\./g, '')
            .replace(/-/g, '')
            .replace(/\s/g, '')
            .toUpperCase();
    }

    /**
     * Separa el RUT en cuerpo y dígito verificador
     * @param {string} rut - RUT a separar
     * @returns {object} - {cuerpo: string, dv: string}
     */
    static separar(rut) {
        const rutLimpio = this.limpiar(rut);
        
        if (rutLimpio.length < 2) {
            return { cuerpo: '', dv: '' };
        }

        return {
            cuerpo: rutLimpio.slice(0, -1),
            dv: rutLimpio.slice(-1).toUpperCase()
        };
    }

    /**
     * Calcula el dígito verificador de un RUT
     * @param {string} cuerpo - Cuerpo del RUT (sin DV)
     * @returns {string} - Dígito verificador calculado
     */
    static calcularDV(cuerpo) {
        const cuerpoLimpio = cuerpo.replace(/\D/g, '');
        
        if (!cuerpoLimpio || cuerpoLimpio.length === 0) {
            return '';
        }

        let suma = 0;
        let multiplicador = 2;

        // Recorrer el cuerpo de derecha a izquierda
        for (let i = cuerpoLimpio.length - 1; i >= 0; i--) {
            suma += parseInt(cuerpoLimpio.charAt(i)) * multiplicador;
            multiplicador = multiplicador === 7 ? 2 : multiplicador + 1;
        }

        const resto = suma % 11;
        const dvCalculado = 11 - resto;

        // Retornar el DV correspondiente
        if (dvCalculado === 11) return '0';
        if (dvCalculado === 10) return 'K';
        return dvCalculado.toString();
    }

    /**
     * Valida si un RUT es correcto
     * @param {string} rut - RUT completo a validar
     * @returns {boolean} - true si es válido
     */
    static validar(rut) {
        const rutLimpio = this.limpiar(rut);

        // Validar longitud mínima
        if (rutLimpio.length < 2) {
            return false;
        }

        // Validar formato (solo números y K/k al final)
        if (!/^\d{7,8}[0-9Kk]$/.test(rutLimpio)) {
            return false;
        }

        const { cuerpo, dv } = this.separar(rutLimpio);
        const dvCalculado = this.calcularDV(cuerpo);

        return dv === dvCalculado;
    }

    /**
     * Formatea un RUT con puntos y guión
     * @param {string} rut - RUT a formatear
     * @returns {string} - RUT formateado (ej: 12.345.678-9)
     */
    static formatear(rut) {
        const { cuerpo, dv } = this.separar(rut);

        if (!cuerpo || !dv) {
            return rut;
        }

        // Agregar puntos cada 3 dígitos
        let cuerpoFormateado = '';
        let contador = 0;

        for (let i = cuerpo.length - 1; i >= 0; i--) {
            if (contador === 3) {
                cuerpoFormateado = '.' + cuerpoFormateado;
                contador = 0;
            }
            cuerpoFormateado = cuerpo.charAt(i) + cuerpoFormateado;
            contador++;
        }

        return `${cuerpoFormateado}-${dv}`;
    }

    /**
     * Valida y muestra mensaje de error en un input
     * @param {HTMLInputElement} input - Input del RUT
     * @param {HTMLElement} errorElement - Elemento para mostrar error (opcional)
     * @returns {boolean} - true si es válido
     */
    static validarInput(input, errorElement = null) {
        const rut = input.value;
        const esValido = this.validar(rut);

        // Agregar o quitar clases de Bootstrap
        if (esValido) {
            input.classList.remove('is-invalid');
            input.classList.add('is-valid');
            if (errorElement) {
                errorElement.textContent = '';
                errorElement.style.display = 'none';
            }
        } else {
            input.classList.remove('is-valid');
            input.classList.add('is-invalid');
            if (errorElement) {
                errorElement.textContent = '❌ RUT inválido. Formato: 12345678-9';
                errorElement.style.display = 'block';
            }
        }

        return esValido;
    }

    /**
     * Auto-formatea el RUT mientras se escribe
     * @param {HTMLInputElement} input - Input del RUT
     */
    static autoFormatear(input) {
        input.addEventListener('input', function(e) {
            const cursorPos = this.selectionStart;
            const valorAnterior = this.value;
            
            // Formatear el RUT
            const rutFormateado = RutValidator.formatear(this.value);
            this.value = rutFormateado;

            // Ajustar posición del cursor
            if (rutFormateado.length > valorAnterior.length) {
                this.setSelectionRange(cursorPos + 1, cursorPos + 1);
            } else {
                this.setSelectionRange(cursorPos, cursorPos);
            }
        });
    }

    /**
     * Genera un RUT aleatorio válido (útil para testing)
     * @returns {string} - RUT válido aleatorio
     */
    static generarAleatorio() {
        const cuerpo = Math.floor(Math.random() * (99999999 - 1000000 + 1)) + 1000000;
        const dv = this.calcularDV(cuerpo.toString());
        return this.formatear(cuerpo.toString() + dv);
    }
}

/**
 * FUNCIONES DE INICIALIZACIÓN
 */

/**
 * Inicializa validación en todos los inputs de RUT de la página
 */
function inicializarValidacionRUT() {
    // Buscar todos los inputs con ID o name que contenga 'rut'
    const rutInputs = document.querySelectorAll('input[id*="rut"], input[name*="rut"], input[id*="Rut"], input[name*="Rut"]');

    rutInputs.forEach(input => {
        // Auto-formatear mientras escribe
        RutValidator.autoFormatear(input);

        // Buscar o crear elemento de error
        let errorElement = input.parentElement.querySelector('.invalid-feedback');
        if (!errorElement) {
            errorElement = document.createElement('div');
            errorElement.className = 'invalid-feedback';
            input.parentElement.appendChild(errorElement);
        }

        // Validar al perder el foco
        input.addEventListener('blur', function() {
            RutValidator.validarInput(this, errorElement);
        });

        // Validar al enviar formulario
        const form = input.closest('form');
        if (form) {
            form.addEventListener('submit', function(e) {
                if (!RutValidator.validarInput(input, errorElement)) {
                    e.preventDefault();
                    input.focus();
                    alert('❌ Por favor corrija el RUT antes de continuar.');
                    return false;
                }
            });
        }
    });

    console.log(`✅ Validación de RUT inicializada en ${rutInputs.length} campo(s)`);
}

/**
 * Separar RUT en dos campos (cuerpo y DV)
 * Útil para formularios que requieren campos separados
 */
function separarRUTenCampos(inputRut, inputCuerpo, inputDV) {
    inputRut.addEventListener('input', function() {
        const { cuerpo, dv } = RutValidator.separar(this.value);
        inputCuerpo.value = cuerpo;
        inputDV.value = dv;
    });
}

// Inicializar cuando el DOM esté listo
if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', inicializarValidacionRUT);
} else {
    inicializarValidacionRUT();
}

// Exportar para uso global
window.RutValidator = RutValidator;
window.inicializarValidacionRUT = inicializarValidacionRUT;
window.separarRUTenCampos = separarRUTenCampos;
